### ISSUE TITLE
Reset password user api

### ISSUE THREAD

**Issue by germanbisogno** [parse-community/parse-server#6817]
**Is your feature request related to a problem? Please describe.**
I'm fan of parse server and after playing around this great tool I found a posible case of improvement.  Currently the implementation of resetting password is only available for the user using a simple form and cannot be done directly from an API.  Even the error messages are not localized and there aren't error codes since it works currently making a post to a single endpoint called /request_password_reset.   So, no way for the developer to handle error codes for a custom implementation of resetting password.

**Describe the solution you'd like**
My proposal is to add an API for resetting the password of the user, something like:

curl -X POST \
  -H "X-Parse-Application-Id: ${APPLICATION_ID}" \
  -H "X-Parse-REST-API-Key: ${REST_API_KEY}" \
  -H "Content-Type: application/json" \
  -d '{"username":"coolguy@iloveapps.com", "token": "token received by email", "new_password": "new password here" }' \
  https://YOUR.PARSE-SERVER.HERE/parse/passwordReset

This will allow to extend the actual implementation and don't depend on just a form to reset the password, by this way the developer will be able to create its own implementation of resetting password from any device or browser.   Additionally the actual feature to use custom pages for choosing the password can still be used in order to customize error validations, localizing messages, etc.

Some benefits that I evaluated:

- Current implementation of resetting password using a form or custom page should continue working.
- By having an API, it extends this feature to be used not only from a browser, but also from a mobile device.

Concerns/Suggestions:

- Is there any concern about security that should be considered on this proposal?  Please, elaborate more, add more audience if necessary.

**Describe alternatives you've considered**
This implementation will require adding new error codes to the [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/parse/ts3.7/index.d.ts) like:

ParseError.USERNAME_NOT_FOUND = 603;
ParseError.RESET_PASSWORD_ERROR = 604;
ParseError.RESET_LINK_EXPIRED = 605;
ParseError.PASSWORD_POLICY_USERNAME = 606;
ParseError.PASSWORD_POLICY_REPEAT = 607;
ParseError.PASSWORD_POLICY_NOT_MEET = 608;

**Additional context**
I'll be providing a possible patch for this implementation to put it in discussion.  This is my first time contributing to this great project and I'll appreciate if this proposal can be considered for future releases of parse server.

Related features:

- Another feature that I considered that could be implemented in a separated ticket can be supporting password less authentication, I think it has been discussed already [here ](https://github.com/parse-community/parse-server/issues/6152)

Thanks so much!

Awesome project guys!

**Comment by germanbisogno** [comment]
diff --git a/src/Controllers/UserController.js b/src/Controllers/UserController.js
index 959f641..1a83e2d 100644
--- a/src/Controllers/UserController.js
+++ b/src/Controllers/UserController.js
@@ -90,7 +90,10 @@ export class UserController extends AdaptableController {
       )
       .then(results => {
         if (results.length != 1) {
-          throw 'Failed to reset password: username / email / token is invalid';
+          throw new Parse.Error(
+            Parse.Error.RESET_PASSWORD_ERROR,
+            'Failed to reset password: username / email / token is invalid'
+          );
         }
 
         if (
@@ -102,7 +105,10 @@ export class UserController extends AdaptableController {
             expiresDate = new Date(expiresDate.iso);
           }
           if (expiresDate < new Date())
-            throw 'The password reset link has expired';
+            throw new Parse.Error(
+              Parse.Error.RESET_LINK_EXPIRED,
+              'The password reset link has expired'
+            );
         }
 
         return results[0];
@@ -246,10 +252,7 @@ export class UserController extends AdaptableController {
     return this.checkResetTokenValidity(username, token)
       .then(user => updateUserPassword(user.objectId, password, this.config))
       .catch(error => {
-        if (error && error.message) {
-          // in case of Parse.Error, fail with the error message only
-          return Promise.reject(error.message);
-        } else {
+        if (error) {
           return Promise.reject(error);
         }
       });
diff --git a/src/Routers/PublicAPIRouter.js b/src/Routers/PublicAPIRouter.js
index 7453835..8a00f10 100644
--- a/src/Routers/PublicAPIRouter.js
+++ b/src/Routers/PublicAPIRouter.js
@@ -203,7 +203,7 @@ export class PublicAPIRouter extends PromiseRouter {
           username: username,
           token: token,
           id: config.applicationId,
-          error: result.err,
+          error: result.err.message,
           app: config.appName,
         });
 
diff --git a/src/Routers/UsersRouter.js b/src/Routers/UsersRouter.js
index 9d015f9..f3f9587 100644
--- a/src/Routers/UsersRouter.js
+++ b/src/Routers/UsersRouter.js
@@ -361,6 +361,40 @@ export class UsersRouter extends ClassesRouter {
     }
   }
 
+  handlePasswordReset(req) {
+    this._throwOnBadEmailConfig(req);
+
+    const { username, token, new_password } = req.body;
+
+    if (!username) {
+      throw new Parse.Error(
+        Parse.Error.USERNAME_MISSING,
+        'you must provide an username'
+      );
+    }
+    return req.config.database
+      .find('_User', {
+        username: username,
+      })
+      .then(results => {
+        if (!results.length || results.length < 1) {
+          throw new Parse.Error(
+            Parse.Error.USERNAME_NOT_FOUND,
+            `No user found with ${username}`
+          );
+        }
+
+        const userController = req.config.userController;
+        return userController
+          .updatePassword(username, token, new_password)
+          .then(() => {
+            return {
+              response: {},
+            };
+          });
+      });
+  }
+
   handleResetRequest(req) {
     this._throwOnBadEmailConfig(req);
 
@@ -473,6 +507,9 @@ export class UsersRouter extends ClassesRouter {
     this.route('POST', '/requestPasswordReset', req => {
       return this.handleResetRequest(req);
     });
+    this.route('POST', '/passwordReset', req => {
+      return this.handlePasswordReset(req);
+    });
     this.route('POST', '/verificationEmailRequest', req => {
       return this.handleVerificationEmailRequest(req);
     });

**Comment by mtrezza** [comment]
Thanks for your suggestion. Feel free to open a PR and we'll be happy to review it.

I suggest you take another look at the current error list and see if you can reuse any of these existing error codes. Also, you may want to be more specific with error codes than `RESET_PASSWORD_ERROR` or use a generic existing error instead.

**Comment by germanbisogno** [comment]
many thanks for your feedback @mtrezza! I'll take a look at your suggestions.

**Comment by germanbisogno** [comment]
hi, 

I just created this [PR](https://github.com/parse-community/parse-server/pull/6830) which is WIP.  Please, could someone with permissions can link this issue to the PR, I'll appreciate.  I created unit tests that will serve as validation for the changes made, please, run them and check comments and failures since currently there is no implementation for new error codes.  I tried to look for existing codes but none fits and for the ones that don't have code yet, I think it is worth to add them to the system.

About your second suggestion @mtrezza "I suggest you think the whole process through and post a step-by-step user flow here. For example, how does the user receive the reset token?" The current flow for resetting password won't change, it is still the same, this will be an addition of an API for resetting password.

Please, feel free to contribute to this PR, I'm exited to see if someone else is interested in this feature and can add a contrubution like adjusting/adding new error codes, checking unit tests, testing, etc.

Thanks!
Ger

**Comment by mtrezza** [comment]
> The current flow for resetting password won't change, it is still the same, this will be an addition of an API for resetting password.

Then how does the user receive the token?

**Comment by germanbisogno** [comment]
hi @mtrezza , sure, I'll create an example.  Just to understand the context, when the user requests the password reset as it is currently implemented in Parse, an email will be sent with a link with the url, if the server is configured to use customPages as the example shows:

  customPages: {
    passwordResetSuccess: <whatever custom page you want>+ "/passwordResetSuccess", <-- your custom page, not necessary needs to be in parse.
    choosePassword: <whatever custom page you want> + "/choosePassword", <-- your custom page, not necessary needs to be in parse.
  },

you can customize the handling of the form for reseting the password, not need to use the default implementation since when the request is done, it sents as parameters in the url username and token, these parameters can be sent later from a custom page using a xhr call to the API in order to confirm the change of password sending the corresponding username, token and new_password.

Thanks, 
Ger

**Comment by mtrezza** [comment]
Let's see if I follow, the current flow is:

1. User requests password -> app calls `User.requestPasswordReset`
1. User receives email with reset link
1. User opens link and gets to page `customPages.choosePassword`
1. User enters new password -> form posted to `/request_password_reset` with user, token, new password
1. Server answers with `302` and redirects to page `customPages.passwordResetSuccess`

This should already be possible with AJAX to skip any redirection, see https://github.com/parse-community/parse-server/pull/5332.

What would be the difference in this PR?

**Comment by germanbisogno** [comment]
you are right @mtrezza , currently it is possible to make ajax requests to the /request_password_reset, but the different errors validations will be thrown with the same code, let's say 

`{code: -1, error: "Failed to reset password: username / email / token is invalid"}`
`{"code":-1,"error":"Password does not meet the Password Policy requirements."}`

etc, so if it is wanted to handle errors from a custom page, won't be possible since the code is alwayts -1 see https://github.com/parse-community/parse-server/blob/142eaa71bdc5cab806706db1c9bbffe9ed47c4d5/src/Routers/PublicAPIRouter.js#L218, probably it is worth to close this PR and create a new one to handle error codes in a different way.

Feel free to close it if doesn't apply.  

Thanks for all your great support on this.
Ger

**Comment by mtrezza** [comment]
If I understand correctly, this PR came with 2 intentions:

>- By having an API, it extends this feature to be used not only from a browser, but also from a mobile device.
>- Parse errors with corresponding code will be used instead of only string messages

Does that mean that:
- your intended scope of the first point is already covered by https://github.com/parse-community/parse-server/pull/5332?
- your intended scope of the second point is to respond with dedicated (new) Parse error codes instead of currently one generic error code and the specific error code in the body?

**Comment by germanbisogno** [comment]
yes that is correct, I thought /request_password_reset endpoint was only intended to be used only from a html page and not as an API, but since you mentioned it, I see that it can receive xhr calls too, so not necessary I think to add another api for the same purpose.  About the error codes yes, in order to handle errors code in a custom page (let's say I want to traslate these codes to another language) it will be better to have specific error codes for each exception, so that exception can be shown in a browser, device, etc.

thanks Manuel
Ger

**Comment by mtrezza** [comment]
> About the error codes yes, in order to handle errors code in a custom page (let's say I want to traslate these codes to another language) it will be better to have specific error codes for each exception, so that exception can be shown in a browser, device, etc.

I agree that having specific error codes would be nicer than the current generic error code where a text has to be parsed on the client side. It would also be more correct, because the current error code `-1: An unknown error or an error unrelated to Parse occurred.` does not seem to apply because the underlying error in known and it is related to Parse.

Do you want to change the PR to introduce these new error codes?

We would not need a new endpoint for this, but it would be a breaking change in Parse Server (for the better), so it would be important that:
- The docs are updated accordingly with the new error codes (try to use existing ones where possible though)
- The change log has to include a highlighted note that this is a breaking change. You can add a draft entry to the change log yourself and look at other entries how breaking changes are highlighted.

**Comment by germanbisogno** [comment]
sounds great Manuel!  let's do that we can change the scope of this PR to improve error codes instead.

many Thanks

Ger

**Comment by mtrezza** [comment]
I already renamed the issue title, you can just change the code and description of the existing PR for improving the error codes.

**Comment by alisaffari97** [comment]
Is it still open to work on it?

**Comment by davimacedo** [comment]
Yes. It is. Feel free to open a PR and let us know if you need any help.
